### ISSUE TITLE
loadParsedData of Profile class does not update Q_max

### ISSUE THREAD

**Issue by rjkoch** [diffpy/diffpy.srfit#59]
[test_1.txt](https://github.com/diffpy/diffpy.srfit/files/3703272/test_1.txt)
[test_2.txt](https://github.com/diffpy/diffpy.srfit/files/3703273/test_2.txt)

If I call this code with the appropriate PDF data file containing qmax metadata

```
from diffpy.srfit.fitbase import FitContribution, FitRecipe, Profile
from diffpy.srfit.pdf import PDFParser, PDFGenerator
import numpy as np

profile = Profile()
parser = PDFParser()
parser.parseFile("test_1.txt")
profile.loadParsedData(parser)

generator = PDFGenerator("G1")

contribution = FitContribution("crystal")
contribution.addProfileGenerator(generator)
contribution.setProfile(profile, xname="r")

recipe = FitRecipe()

recipe.addContribution(contribution)
```

and then call `recipe.crystal.G1.getQmax()` I get the correct qmax from the file metadata.

If I then however replace the profile data with
```
parser = PDFParser()

parser.parseFile("test_2.txt")

metadata = parser.getMetaData()

file_q_max = metadata['qmax']

recipe.crystal.profile.loadParsedData(parser)

new_q_max = recipe.crystal.G1.getQmax()
```

`new_q_max ` will still contain the old qmax value, from the first file.

This will successfully change it:
```
parser = PDFParser()

parser.parseFile("test_2.txt")


metadata = parser.getMetaData()

file_q_max = metadata['qmax']

recipe.crystal.profile.loadParsedData(parser)

new_q_max = recipe.crystal.G1.getQmax()
    
recipe.crystal.G1.setQmax(file_q_max)
```

Is this the expected behavior? Should I be doing this a different way? Perhaps instantiating a new  `Profile` and passing that to the recipe through `recipe.crystal.setProfile`?

Below is the full code for a working example. Data is attached
```
from diffpy.srfit.fitbase import FitContribution, FitRecipe, Profile
from diffpy.srfit.pdf import PDFParser, PDFGenerator
import numpy as np

def check_qmax(desired,actual):
    print(f"\nMetadata in new file suggests Q_max should be {desired}, q_max is now {actual}")
    if not np.isclose(actual,desired):
        print("\nThese are not equal\n")
    elif np.isclose(actual,desired):
        print("\nThese are equal\n")
    

profile = Profile()
parser = PDFParser()
parser.parseFile("test_1.txt")
profile.loadParsedData(parser)

generator = PDFGenerator("G1")

contribution = FitContribution("crystal")
contribution.addProfileGenerator(generator)
contribution.setProfile(profile, xname="r")

recipe = FitRecipe()

recipe.addContribution(contribution)

old_q_max = recipe.crystal.G1.getQmax()


parser = PDFParser()

parser.parseFile("test_2.txt")


metadata = parser.getMetaData()

file_q_max = metadata['qmax']

recipe.crystal.profile.loadParsedData(parser)

new_q_max = recipe.crystal.G1.getQmax()

check_qmax(file_q_max,new_q_max)
    
recipe.crystal.G1.setQmax(file_q_max)

new_q_max = recipe.crystal.G1.getQmax()
check_qmax(file_q_max,new_q_max)
```

**Comment by stevenhua0320** [comment]
@sbillinge This issue is about whether we should keep the clear distinction for the functionality of `getter` and `setter`. Currently in our code, if we have two pieces of data, if we parse the first set of data, load it into the `Profile` object, and set it into the `FitRecipe`, then everthing for the attributes would be set to corresponding to the first set of data; Then when we parse the second set of data, before we reset the `Fitrecipe` object with the second set of data, we load the `Qmax` of the recipe, and now it would return the `Qmax` for the first set of data. Only if we set the `Profile` for the second piece of data, then the `Qmax` would be changed to the second piece of dataset. Current behavior makes sense to me, but for some users who do not familar with the `getter` and `setter` functionality it might be a little bit hard to adapt like what it says here. So two possible approaches to address this issue here:
1. We still keep the current behavior of the code since it could keep the clear behavior of the `getter` and `setter` in the functionality perspective of the call of the method. It is also easier for the developer to handle the issue when in the future a bug arises when some issues related to the use of `getter` and `setter`. What we need to do is possibly need to include the example here in the documentation to instruct the users how to set the data every time before doing anything else. 
2. Do a refactor of the code, when we have a piece of data in the `FitRecipe` and we have another piece of dataset, when we parse the dataset and at the load step, it loads and covers the `Fitrecipe` with second piece of data in replace of the first set of data. It would be (somewhat) easier for the users, but there might be higher possibility to have a bug/unintended behavior since there might be other attributes set in the `FitRecipe` with the first piece of data that we should also change, which would distort the result. Moreover, it might be harder for the future developer to do the maintainance since in this case the `getter` and `setter` is intertwined in the method call. 

What do you think?

**Comment by sbillinge** [comment]
Yes. This is a tricky question. The code is behaving magically from the perspective of the user (I loaded new data. Why the data updated but the qmax didn't?) but the code is logical and not magical as you mention.  Let's discuss this a bit and think about it.

**Comment by sbillinge** [comment]
One thought is that, If there is a way for s profile to know about any recipes it is a part of, to have a step where the new profile is reloaded?  Is there a way for this to happen?

**Comment by stevenhua0320** [comment]
Right now we don't have such way. We could actually go into high level about the use case here. 
The first use case here is, given that we have all experimental data and the required config (`Qmax`, `Qdamp`, etc.) Now we want to fit the data with the same recipe and data. What marks the difference is that we change the `Qmax` parameter to the second piece of data after parsing and loading all the second piece of data. What the user needs is not about loading the second piece of data at all, it is just updating the `Qmax`. So, in this case, we only need to reset the `Qmax` parameter in the recipe, which we already have `recipe.crystal.G1.setQmax(...)`

But we can also have another use case, which is after we have done on all the work on the first piece of data, then we have another set of data, we parse and load it, and then in this case we prefer to fit it in a new `FitRecipe` object since everything including the parameter and the data would change. If we insist on the same `FitRecipe` object I would think it as a bad case since user might want expect to fit the new data with new parameters, but in actual they ends up having confusion whether they use the correct `parameter`/ `dataset` as they want.

**Comment by sbillinge** [comment]
I read rob's issue a bit more in detail, and i think the code is behaving as it should in terms of the setters and getters.

The slight "magic" is that, if I type `recipe.crystal.profile.loadParsedData(parser)`, naively I might expect that it loads the the parsed-data correctly, which means that it correctly loads the new `Q_max` which is a legitimate attribute of the data.  

Could you look quickly into the code in `load_parsed_data` and see if there is a simple fix by runninge the qmax setter somehow?  If there is a quick fix we can do it, and if it is a big refactor we can push this off.

**Comment by rjkoch** [comment]
Happy to see this is still active, even if I am not!

**Comment by sbillinge** [comment]
Great to hear from you Rob. I hope things are well with you!

**Comment by stevenhua0320** [comment]
@sbillinge I think the process of how current code works now is first we have a `Profile` object, we load the dataset into the `Profile`. Then we need to pass the metadata (generators including `Qmax`) into the `Generator` instance. Then we use the `Recipe` to produce the result based on the metadata and the dataset. Even though when we run `loadParsedData` would correctly load the data, it would not directly be updated since we must pass these metadata to the `Generator` so that it could identify the new metadata and do the update. Additionally, it would not only update the `Qmax` but also all the metadata here when we run the `setProfile`. However, things would be very tricky here, in the issue described here this makes sense as the test data files given, we are investigating the same material and the only difference in the metadata is the `Qmax` numerical difference(although we also have different observed datasets below). If we have another case that we are investigating the same material with all metadata setup are different (`Qdamp`, `Qbroad` etc.), then it would be undesireable behavior if we only update one parameter here.

**Comment by sbillinge** [comment]
I am not sure I agree.  Qmax is an attribute of the data that resides in the input file (in general).  The other parameters that you mention are fitting parameters (sometimes fixed) so they have a very different flavor.  I think it would be fine, desirable even, to update Qmax without updating any other parameters.
